For Ubuntu 18.04, Create a new snippet for Netplan which supports multiple interfaces

stefanlasiewski opened the pull request:
1. Create a new snippet to configure a network using Netplan.io. Unlike #545, this template should support additional interfaces within `/etc/netplan/01-netcfg.yaml`.
2. In the preseed_default_finish, create a conditional and use the snippet if the OS is Ubuntu
18.04, or use the old snippet if it's an older version of Ubuntu. This borrows from the comment at https://github.com/theforeman/community-templates/pull/545#issuecomment-454434975

This is similar to the work done by @pacaj2am for PR #545, and borrows some code from his contribution. This template borrows heavily from the original `preseed_networking_setup.erb`.

stefanlasiewski committed:
1. Create a new Snippet to configure a network using Netplan.io
2. In the preseed_default_finish, use the Snippet if the OS is Ubuntu
18.04

Kudos to @pacaj2am & PR #545 for the inspiration.

stefanlasiewski commented:
This PR is incomplete. I actually could use some help figuring out what I'm missing.

When I try this, Foreman throws the following error. I'm not sure which instance of `gateway` this error is referring to. For non-primary interfaces, I'm confused about how to refer to certain elements like the gateway, netmask, etc.

> Warning! There was an error rendering the Preseed default finish template: ERF42-7327 [Foreman::Exception]: The snippet 'preseed_networking_netplan_setup' threw an error: undefined method '#gateway' for NilClass::Jail (NilClass)

ekohl commented on provisioning_templates/finish/preseed_default_finish.erb:
```diff
@@ -53,5 +56,10 @@ oses:
-<%= snippet 'preseed_networking_setup' %>
+<%# Ubuntu 18.04 and newer uses netplan instead of /etc/network/interfaces -%>
+<% if is_ubuntu && os_major >= 18 -%>
+  <%= snippit 'preseed_networking_netplan_setup' %>
```
```suggestion
  <%= snippet 'preseed_networking_netplan_setup' %>
```

stefanlasiewski replied:
Good catch. Thanks.

ekohl commented on provisioning_templates/snippet/preseed_networking_netplan_setup.erb:
```diff
@@ -0,0 +1,62 @@
+<% else -%>
+      addresses: [ <%= @host.ip %>/<%= host_subnet.cidr %><% if @host.ip6 && host_subnet6 && !host_dhcp6 -%>, <%= @host.ip6 %>/<%= host_subnet6.cidr %><% end -%> ]
+      gateway4: <%= host_subnet.gateway %>
+      <% if host_subnet6.gateway -%>gateway6: <%= host_subnet6.gateway %><% end %>
```
You probably need to check if `host_subnet6` exists before using it. That's the likely source

stefanlasiewski replied:
Great catch. Thanks.

ekohl commented on provisioning_templates/snippet/preseed_networking_netplan_setup.erb:
```diff
@@ -0,0 +1,62 @@
+      dhcp4: true
+<% else -%>
+      addresses: [ <%= @host.ip %>/<%= host_subnet.cidr %><% if @host.ip6 && host_subnet6 && !host_dhcp6 -%>, <%= @host.ip6 %>/<%= host_subnet6.cidr %><% end -%> ]
+      gateway4: <%= host_subnet.gateway %>
```
Could be good to be defensive here and also consider the case where a IPv4 subnet is not present.

stefanlasiewski replied:
👍

stefanlasiewski committed:
Fix typo with 'snippit'

stefanlasiewski committed:
Test that `host_subnet6` exists before using it.

Resolves https://github.com/theforeman/community-templates/pull/568/files/2bf175c92586c3d1c7cbabb0b30385f88b52ec13#r270358453

stefanlasiewski committed:
Test that `host_subnet6` exists before using it.

Resolves https://github.com/theforeman/community-templates/pull/568/files/2bf175c92586c3d1c7cbabb0b30385f88b52ec13#r270358453

stefanlasiewski committed:
Merge branch 'netplan-2' of https://github.com/stefanlasiewski/community-templates into netplan-2

stefanlasiewski committed:
Test if IPv4 IP & subnet exists before printing things.

stefanlasiewski commented on provisioning_templates/snippet/preseed_networking_netplan_setup.erb:
```diff
@@ -0,0 +1,62 @@
+<% if host_dhcp -%>
+      dhcp4: true
+<% else -%>
+      addresses: [ <%= @host.ip %>/<%= host_subnet.cidr %><% if @host.ip6 && host_subnet6 && !host_dhcp6 -%>, <%= @host.ip6 %>/<%= host_subnet6.cidr %><% end -%> ]
```
Since an interface may have multiple IPv4 and/or multiple IPv6 Ips in one set, it might make sense to loop here. A format like the following might make looping easier:
```
addresses:
  - 1.2.3.4/24
  - 5.6.7.8/21
```

Not sure I'm up for that right now. My ERB skills are pretty basic right now.

ekohl replied:
It'd probably be something like
```erb
<%-
addresses = []
addresses << "#{@host.ip}/#{host_subnet.cidr}" if @host.ip && host_subnet && !host_dhcp
addresses << "#{@host.ip6}/#{host_subnet6.cidr}" if @host.ip6 && host_subnet6 && !host_dhcp6
-%>
```

Then later
```erb
<% unless addresses.empty? -%>
addresses:
<% addresses.each do |address| -%>
  - <%= address %>
<% end -%>
<% end -%>
```

stefanlasiewski commented on provisioning_templates/snippet/preseed_networking_netplan_setup.erb:
```diff
@@ -0,0 +1,62 @@
+<% if interface_dhcp -%>
+      dhcp4: true
+<% else -%> 
+      addresses: [ <%= interface.ip %>/<%= interface_subnet.cidr %><% if interface.ip6 && interface_subnet6 && !interface_dhcp6 -%>, <%= interface.ip6 %>/<%= interface_subnet6.cidr %><% end -%> ]
```
As I mentioned above, it might make sense to loop through Ips here.

stefanlasiewski committed:
Add another conditional to skip code for /etc/resolv.conf and
/etc/network/interfaces

stefanlasiewski committed:
Merge branch 'develop' into netplan-2

Elyytscha reviewed (approved):
can approve that those templates work for ubuntu 18.04 with two interfaces with different subnets

Elyytscha commented on provisioning_templates/snippet/preseed_networking_netplan_setup.erb:
```diff
@@ -0,0 +1,62 @@
+      addresses: [ <%= interface.ip %>/<%= interface_subnet.cidr %><% if interface.ip6 && interface_subnet6 && !interface_dhcp6 -%>, <%= interface.ip6 %>/<%= interface_subnet6.cidr %><% end -%> ]
+      gateway4: <%= interface_subnet.gateway %>
+        nameservers:
+          search: <%= @interface.domain %>
```
maybe here should be a check if a domain is set for the interface, because if no domain is set or the org // location can't access it, it throws an error

Elyytscha commented on provisioning_templates/snippet/preseed_networking_netplan_setup.erb:
```diff
@@ -0,0 +1,62 @@
+      dhcp4: true
+<% else -%> 
+      addresses: [ <%= interface.ip %>/<%= interface_subnet.cidr %><% if interface.ip6 && interface_subnet6 && !interface_dhcp6 -%>, <%= interface.ip6 %>/<%= interface_subnet6.cidr %><% end -%> ]
+      gateway4: <%= interface_subnet.gateway %>
```
here you will end up with two default gateways in case of two interfaces which will break the network functionallity

tbrisker commented:
ping @stefanlasiewski - what's the status here? looks like it needs a rebase and some comments need addressing?

tbrisker commented:
On behalf of the Foreman team, thank you for your contribution to the Foreman community templates repository! 
This repository is being archived and all templates are migrated to the relevant repositories. If you wish to pursue this change, please re-open the PR against the relevant repository. For more information and background, please [click here]( https://community.theforeman.org/t/heads-up-community-templates-repo-will-be-archived-tomorrow/19539).
